### PR TITLE
Support dynamic locales

### PR DESCRIPTION
**Pull request by lieut-data**
This tweaks the locale support in a couple of ways. First, we now locate locales in the root `assets/i18n` directory, resolved via `GetBundlePath`. This hardens the plugin against changes the server might make it how it unpacks the plugin, and simplifies the distribution of i18n resources for users.

Secondly, we now register all locales found in the `assets/i18n` directory. This means that adding support for new locales simply requires dropping in the file without any need to modify the code or even recompile.

Learned a bunch about the framework as part of making these changes, so thanks for the opportunity :)

### CODE REVIEW

**Comment by lieut-data** [comment]
Perusing the code some more, I realize there are a bunch of `func *En` methods that perhaps have English assumptions built into them. Though I note they all use translated strings, and the non-`en` locales fall back to them.

Feel free to push back if the intent was to prevent other locales since it wouldn't really work!

File: server/i18n.go [hunk 1]
```diff
@@ -4,42 +4,45 @@
...
 )
 
-var locales = make(map[string]string)
-
 func (p *Plugin) TranslationsPreInit() error {
-
-	i18nDirectory, found := fileutils.FindDir(*p.ServerConfig.PluginSettings.Directory + "/" + manifest.Id + "/server/dist/i18n/")
-	if !found {
-		return fmt.Errorf("unable to find i18n directory")
+	bundlePath, err := p.API.GetBundlePath()
```
Comment (scottleedavis) [inline server/i18n.go]: awesome!

File: server/i18n.go [hunk 2]
```diff
@@ -4,42 +4,45 @@
...
-		if filepath.Ext(f.Name()) == ".json" {
-			filename := f.Name()
-			locales[strings.Split(filename, ".")[0]] = filepath.Join(i18nDirectory, filename)
-
+		filename := f.Name()
+		if filepath.Ext(filename) == ".json" {
 			if err := i18n.LoadTranslationFile(filepath.Join(i18nDirectory, filename)); err != nil {
-				return err
+				p.API.LogError("Failed to load translation file", "filename", filename, "err", err.Error())
+				continue
```
Comment (scottleedavis) [inline server/i18n.go]: better continuing here and logging, nice.

File: server/i18n_test.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/mattermost/mattermost-server/plugin/plugintest"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTranslationsPreInit(t *testing.T) {
```
Comment (scottleedavis) [inline server/i18n_test.go]: thank you for adding test coverage on this.  I learned some things from your style.

**Comment by scottleedavis** [comment]
> func *En

I believe you are referring to funcs like `createOccurrencesEN` etc, which are meant to be english specific (parsing grammar strings), and will need specific handling for each supported local beyond the i18n json file.  

e.g. 
```go
func (p *Plugin) CreateOccurrences(request *ReminderRequest) error {

	user, uErr := p.API.GetUserByUsername(request.Username)
	if uErr != nil {
		return uErr
	}
	_, locale := p.translation(user)

	switch locale {
	case "en":
		return p.createOccurrencesEN(request)
	default:
		return p.createOccurrencesEN(request)
	}

}
```

The only reason this is done because of the differences in ordering of grammar between languages.  I am hoping to start spanish soon.

**Comment by scottleedavis** [comment]
Tested locally and works great, thank you @lieut-data !

**Pull request merged**